src/openeo_aggregator/caching.py

```diff
@@ -0,0 +1,446 @@
+        callback = self._wrap_logging(callback=callback, key=key)
+
+        with self._zk_connect_or_not() as connected_client:
+
+            def handle(found: Any = UNSET, store: str = None, debug=None, error=None) -> Any:
+                """Helper to compactly handle various cache miss/hit/error situations"""
+                if error:
+                    _log.error(f"{self!r} {error}")
+                elif debug:
+                    _log.debug(f"{self!r} {debug}")
+
+                if found is not UNSET:
+                    value = found
+                else:
+                    value = callback()
+                    try:
+                        # Try to store value (but skip any failure)
+                        if store == "create":
+                            connected_client.create(path=path, value=self._serde.serialize(value), makepath=True)
+                        elif store == "set":
+                            connected_client.set(path=path, value=self._serde.serialize(value))
+                    except kazoo.exceptions.NodeExistsError:
+                        # When creating node that already exists: another worker probably set cache in the meantime
+                        _log.warning(f"{self!r} failed to create node {path!r}: already exists.")
+                    except Exception as e:
+                        _log.error(f"{self!r} failed to {store} {path!r}: {e!r}")
+
+                return value
+
+            if connected_client is None:
+                return handle(store=None, error="no connection")
+            # Let's see if we find something in cache.
+            try:
+                zk_value, zk_stat = connected_client.get(path=path)
+            except kazoo.exceptions.NoNodeError:
+                # Simple cache miss: no zookeeper node, so we'll have to create it.
+                return handle(store="create", debug=f"cache miss {path!r}")
+            except Exception as e:
+                return handle(store=None, error=f"unexpected get failure: {e!r}")
+
+            # We found something, check expiry and validity
+            if zk_stat.last_modified < self._valid_threshold:
+                return handle(store="set", debug=f"invalidated {path!r}")
+            if zk_stat.last_modified + ttl <= Clock.time():
+                # Note that we evaluate the TTL at "get time",
+                # instead of storing an expiry threshold at "set time" along with the data
+                return handle(store="set", debug=f"expired {path!r}")
+            try:
+                # Can we read it?
+                value = self._serde.deserialize(zk_value)
+            except json.JSONDecodeError as e:
+                # Cache hit, but corrupt data: update it
+                return handle(store="set", error=f"corrupt data on {path!r}: {e!r}")
+
+            return handle(found=value, store=None, debug=f"cache set {path!r}")
+
+    def invalidate(self):
+        self._valid_threshold = Clock.time()
```

Comment:
This invalidates just for current process, but other workers might not be aware of invalidation: instead remove while zk subtree?